### ISSUE TITLE
Add some way to tell if the JVM is launched with the option to checkpoint

### ISSUE THREAD

**Issue by tjwatson** [CRaC/org.crac#12]
For some scenarios it would be best if a resource could delay any action until the restore side.  With the current org.crac (and jdk.crac) APIs it does not seem possible for a Resource to avoid work before a checkpoint.  The resource always has to assume there will be no checkpoint in the future and do any work necessary to get the resource functional.  Then it must prepare for the checkpoint with the `beforeCheckpoint`, possibly undoing what it has done so far, close files etc.  Then it must recover with the `afterRestore` method to get back in a good state for the restore process.

It would be nice if a resource could avoid all work on the checkpoint side and only be concerned about getting into a working state on the restore side with the `afterRestore` method.  Perhaps some method on the `Core` like `boolean isCheckpointSupported()` would work.  For example, on a jdk.crac implementation this method would return true if the `-XX:CRaCCheckpointTo` JVM option was used.  Otherwise it would return false.

**Comment by tjwatson** [comment]
Another option could be to have a method on `Context` that runs a lambda on restore like this:

```java
    /**
     * A function to call on restore
     *
     * @param <T> Exception thrown by the call method
     */
    @FunctionalInterface
    public static interface OnRestore<T extends Throwable> {
        /**
         * Called on restore
         *
         * @throws T on a restore error. Any exception throw will result in a failure
         *               to restore the process.
         */
        public void call() throws T;
    }

    /**
     * Runs the given function on restore if the runtime has been configured to perform
     * a checkpoint; otherwise the function is run immediately from the calling thread synchronously.
     *
     * @param <T>   The type of throwable the function may throw
     * @param toRun The function to run on restore.
     * @throws T any errors that occur while running the function. If an exception is thrown during
     *               restore then the process restore will fail.
     */
    public <T extends Throwable> void onRestore(OnRestore<T> toRun) throws T;
```

This way a resource that needs to run some logic in "normal" launches synchronously, but delay such logic until restore can simply do this:

```java
Core.getGlobalContext().onRestore(() -> {
   // insert logic here that needs to always run when not doing a checkpoint;
   // but needs to delay until restore when doing a checkpoint
});
```

**Comment by tjwatson** [comment]
The other advantage to this method is that the lambda doesn't need to be weak referenced by the org.crac implementation.  It is either going to be run immediately and de-referenced for GC or it is going to need to be strongly referenced to be run on the restore side.  Once restored it can be de-referenced for GC.

**Comment by AntonKozlov** [comment]
Do I understand correctly that you want to perform some intialization steps either during startup (if no checkpoint is guaranteed), or postpone the steps until restore but not do them during startup in this case? I can imagine a case of re-initialization that is considerably harder than a single forward initialization.

The current design of Resource is intentionally disconnected from the particular timing of checkpoint(s). The Resource, once initalized, in general case should assume multiple, potentially unlimited number of checkpoints upfront. Evertime, CRaC API calls beforeCheckpoint and then afterRestore. In that sense, the API to determ "no more checkpoints are possible" should not only detect not a single checkpoint is possible, but to predict e.g. no more jcmd requests, which is way harder :)

As a less convenient substituion, it is possible to provide the deployment configuration of the app via system properties, like

```sh 
java -jar app.jar
# or
java -XX:CRaCCheckpointTo=dir -Dapp.checkpoint.count=1 -jar app.jar
```

And the code
```java

public class R implements Resource {
    private int checkpointCount = Integer.getInteger("app.checkpoint.count", 0);
    
    R() {
        if (checkpointCount == 0) {
            init();
        }
    }
    
    @Override
    public void beforeCheckpoint(Context<? extends Resource> ctx) { }

    @Override
    public void afterRestore(Context<? extends Resource> ctx) {
	if (--checkpointCount == 0) {
            init();
        }
    }
}
```

Definetely not pretty, but should work.

Overall the proposed API looks useful, but we should to think about names. The `isCheckpointSupported()` can be interpreted as the overall VM CRaC support status, and the way how VM was started. How about something like below?

```java
/**
 * Checks configuration for the possible checkpoint.
 *
 * @return false if the checkpoint is not possible in the current configuration.
 */
public static boolean isCheckpointConfigured();
```

**Comment by tjwatson** [comment]
My usecase is going to typically involve doing a checkpoint during a container image build and then doing a single restore when the container image is started in a k8s environment (AWS, Azure, OpenShift etc.).

During such a container image build I do not want the application to make any connections to a database.  I want that to be delayed until restore where the environment is properly configured for the production database (with credentials, datasource location etc.).  This is one example where either the application itself, or some framework the application uses would want to do some operation only after the first restore.  Or in the case where there is no chance of a checkpoint happening, do the connection right at application startup (the "normal" launch case).

A `Resource` is useful for cases where something wants to be initialized always and then tear down its state on checkpoint and re-initialize state on restore.  But here I can imagine usecases that don't want to establish any state until the first restore.  After that, if the usecase supports multiple checkpoint/restore sequences then it would use a `Resource` to do that.



> How about something like below?
> 
> ```java
> /**
>  * Checks configuration for the possible checkpoint.
>  *
>  * @return false if the checkpoint is not possible in the current configuration.
>  */
> public static boolean isCheckpointConfigured();
> ```

`isCheckpointConfigured` is a fine name in my opinion.  Thanks for considering.

**Comment by rvansa** [comment]
Right now the only way to 'enable' checkpoint is to set `-XX:CRaCCheckpointTo=...` but in the future this shouldn't be the only way; it makes a perfect sense to make this manageable. Technically there are some things we do for CRaC when the JVM boots and would slow down the startup by a few milliseconds had we done this on every JVM boot, so we can assume that there would be a `-XX:+EnableCRaC` option. However, the value of `isCheckpointConfigured()` should be considered volatile during runtime.

**Comment by tjwatson** [comment]
> However, the value of isCheckpointConfigured() should be considered volatile during runtime.

Interesting.  So one could consider enabling a "checkpoint" JVM dynamically (one that was not started with `-XX:CRaCCheckpointTo=...`).  That kind of advanced usecase really makes things much more complicated for scenarios that want to avoid doing any kind of work on the "checkpoint" side.

I'm not sure I see the advantages of that advanced behavior to outweigh the complications it would introduce to every other scenario that knows well ahead that the JVM will be used to do a checkpoint.

**Comment by rvansa** [comment]
You, as the author of the application, have a world of options to communicate to the application that it is meant to be checkpointed - system property as Anton shows is a great and simple way. The application that expects only single scenario can expose an internal API like you propose. On the other hand the API should be generic enough to allow other usecases - e.g. when you want to run a stateful application on a spot instance in cloud, and persist it when you receive the termination notice.

One property of this `afterRestore`-only API I dislike is that the execution won't be 'trained' during checkpoint, and if it gets executed only after restore you might be classloading, running in interpreted mode, compiling... for a quick operation right after restore all this should be moved before checkpoint. When I optimize for quick startup, the time for build + boot + checkpoint is considered insignificant, contrary to anything that happens after restore (in production, and repeatedly).

At some point we might even add an option to do 'dry-run checkpoints' automatically - closing all resources and reopening again. It might help even functionally; I know about a usecase where the first attempt to checkpoint fails because during `beforeCheckpoint` the application lazily instantiates a resource; only the second attempt succeeds.
